lib/navigation/preprocess_encoders.cpp

```diff
@@ -1,56 +1,106 @@
+                num_reliable_encoders_,
+                core::kNumEncoders);
+    return std::nullopt;
+  }
+  if (num_reliable_encoders_ == core::kNumEncoders) {
+    auto reliable_data                    = encoder_data;
+    const Quartile quartiles              = getQuartiles(reliable_data);
+    const core::Float interquartile_range = quartiles.q3 - quartiles.q1;
+    return {
+      {.median      = quartiles.median,
+       .upper_bound = quartiles.median + static_cast<core::Float>(1.5) * interquartile_range,
+       .lower_bound = quartiles.median - static_cast<core::Float>(1.5) * interquartile_range}};
+  } else {
+    std::array<uint32_t, core::kNumEncoders - 1> reliable_data;
+    std::size_t j = 0;
+    for (std::size_t i = 0; i < encoder_data.size(); ++i) {
+      if (are_encoders_reliable_.at(i)) {
+        reliable_data.at(j) = encoder_data.at(i);
+        ++j;
+      }
+    }
+    const Quartile quartiles              = getQuartiles(reliable_data);
+    const core::Float interquartile_range = quartiles.q3 - quartiles.q1;
+    return {
+      {.median      = quartiles.median,
+       .upper_bound = quartiles.median + static_cast<core::Float>(1.2) * interquartile_range,
+       .lower_bound = quartiles.median - static_cast<core::Float>(1.2) * interquartile_range}};
+  }
 }
 
-core::EncoderData EncodersPreprocessor::detectOutliers(const core::EncoderData encoder_data)
+std::optional<core::EncoderData> EncodersPreprocessor::sanitise(
+  const core::EncoderData &encoder_data)
 {
-  /*
-  TODOLater: implement
-  rough process:
-  - get q1, median, q3 of encoder array
-  - define upper & lower bounds as (-)1.5*inter-quatrile range
-  - if any datapoint outwith this, set
-  num_outliers_per_encoder_[i] += 1
-  -set outlier points as median
-
-  -also has to be able to handle 1 unreliable sensor
-  -also figure out return type/ what we update and update
-  documentation as appropriate
-  */
-  return {0, 0, 0, 0};
+  const auto statistics = getStatistics(encoder_data);
+  if (!statistics) {
+    logger_.log(core::LogLevel::kFatal, "Failed to obtain statistics for measurement");
+    return std::nullopt;
+  }
+  auto sanitised_data = encoder_data;
+  for (std::size_t i = 0; i < sanitised_data.size(); ++i) {
+    if (sanitised_data.at(i) > statistics->upper_bound
+        || sanitised_data.at(i) < statistics->lower_bound || !are_encoders_reliable_.at(i)) {
+      ++num_consecutive_outliers_per_encoder_.at(i);
+      sanitised_data.at(i) = statistics->median;
+    } else {
+      num_consecutive_outliers_per_encoder_.at(i) = 0;
+    }
+  }
+  if (checkReliable() == SensorChecks::kUnacceptable) { return std::nullopt; }
+  for (std::size_t i = 0; i < are_encoders_reliable_.size(); ++i) {
+    if (!are_encoders_reliable_.at(i)) { sanitised_data.at(i) = statistics->median; }
+  }
+  return sanitised_data;
 }
 
-void EncodersPreprocessor::checkReliable(const core::EncoderData &encoder_data)
+SensorChecks EncodersPreprocessor::checkReliable()
```

Comment:
is there a reason for `SensorChecks` instead of `Result`? this could be argued either way tbh

Reply:
Yeah just the naming is much more appropriate for what this does because we can have a sensor already failed which is "acceptable" but not really a resounding "success"...

Reply:
yea that's fair